src/main/ruby/core/io.rb

```diff
@@ -146,13 +146,30 @@ def unshift(str, start_pos)
     end
 
     def fill(io)
-      Truffle.primitive :iobuffer_fill
-
       unless io.kind_of? IO
-        return fill(io.to_io)
+        io = io.to_io
       end
 
-      raise PrimitiveFailure, 'InternalBuffer#fill primitive failed'
+      max = 8192
+      left = @total - @used
+      count = left < max ? left : max
+
+      begin
+        buffer = Truffle::POSIX.read_string(io.descriptor, count)
+      rescue Errno::EAGAIN
+        retry
```

Comment:
This could infinite loop if the `io` object refers to a non-blocking FD. I haven't traced MRI to see if that's expected behavior. Likewise, should a delay be inserted so it's not just spinning?

Reply:
Looking at it more, it was ported from Java. But the question still stands.

Reply:
Yes, indeed, I was thinking to it as well. I'll add a `select` call there to wait in case IO is in non-blocking mode.